### PR TITLE
fix(npm): use lockfileVersion to determine npm version

### LINKED ISSUE #25817: Renovate doesn't respect package-lock.json lockfileVersion to select NPM version to install
### Discussed in https://github.com/renovatebot/renovate/discussions/25815

<div type='discussions-op-text'>

<sup>Originally posted by **kvanzuijlen** November 15, 2023</sup>
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.59.4

### If you're self-hosting Renovate, select which platform you are using.

Github.com

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the problem

    🚧 Doesn't work for both the GitHub App as for the self-hosted app

Previously, Renovate selected the NPM version to install based on configured constraints or the lockfileVersion it detected. This no longer works. I've set up a quick reproduction repo over here: https://github.com/kvanzuijlen/renovate-reproduction.

The issue this causes shows up in kvanzuijlen/renovate-reproduction#3

The error is as follows:
```log
ERROR: npm v10.2.3 is known not to run on Node.js v14.15.5.  This version of npm supports the following node versions: `^18.17.0 || >=20.5.0`. You can find the latest version at https://nodejs.org/.
```

The lockfileVersion is 1, so it should only be used with NPM v5 or v6 (see below), however, it tries to install NPM 10.2.3 (at the moment of writing latest) instead.

```
1: The lockfile version used by npm v5 and v6.
2: The lockfile version used by npm v7 and v8. Backwards compatible to v1 lockfiles.
3: The lockfile version used by npm v9. Backwards compatible to npm v7.
```

The node version is pinned to v14.15.5 by the .nvmrc file in the repository.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Generating package-lock.json for . (branch="renovate/express-4.x-lockfile")
DEBUG: Spawning npm install to create ./package-lock.json (branch="renovate/express-4.x-lockfile")
DEBUG: Updating lock file only (branch="renovate/express-4.x-lockfile")
DEBUG: Using node constraint "14.15.5" from .nvmrc (branch="renovate/express-4.x-lockfile")
DEBUG: Performing lockfileUpdate (npm) (branch="renovate/express-4.x-lockfile")
DEBUG: Setting CONTAINERBASE_CACHE_DIR to /tmp/worker/1619d6/0379b2/cache/containerbase (branch="renovate/express-4.x-lockfile")
DEBUG: Using docker to execute (branch="renovate/express-4.x-lockfile")
{
  "image": "sidecar"
}

DEBUG: Resolved stable matching version (branch="renovate/express-4.x-lockfile")
{
  "toolName": "npm"
  "constraint": null
  "resolvedVersion": "10.2.3"
}

```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

**Edit:** As far as I can tell this bug seems to have been introduced in #23131
**Edit 2:** I'm willing to work on a fix for this issue myself, but I'm currently at a loss on how I would go about fixing this</div>

### PR DESCRIPTION
**Pull request by rarkins**
## Changes

Use lockfileVersion in package-lock.json to determine npm version if no constraints are previously found

## Context

Fixes #25817

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/npm/post-update/npm.ts [hunk 1]
```diff
@@ -48,7 +82,9 @@ export async function generateLockFile(
       toolName: 'npm',
       constraint:
         config.constraints?.npm ??
-        getPackageManagerVersion('npm', await lazyPgkJson.getValue()),
+        getPackageManagerVersion('npm', await lazyPgkJson.getValue()) ??
```
Comment (kvanzuijlen) [inline lib/modules/manager/npm/post-update/npm.ts:85]: While looking into this myself I noticed a slight typo(?). I think it should be `lazyPkgJson`, correct?

**Review by kvanzuijlen — APPROVED** [review]
LGTM! Thx for the quick fix @rarkins

File: lib/modules/manager/npm/post-update/npm.ts [hunk 2]
```diff
@@ -28,6 +28,40 @@ import { getNodeToolConstraint } from './node-version';
...
+  lockFileDir: string,
+): Promise<string | undefined> {
+  const lockFileName = upath.join(lockFileDir, 'package-lock.json');
+  const lockFile = await readLocalFile(lockFileName, 'utf8');
+  // istanbul ignore if: should not happen
+  if (!lockFile) {
+    return undefined;
+  }
+  try {
+    const lockFileJson = JSON.parse(lockFile);
```
Comment (viceice) [inline lib/modules/manager/npm/post-update/npm.ts:41]: we should use our JSON util to parse

Comment (rarkins) [reply lib/modules/manager/npm/post-update/npm.ts:41]: Which one? We don't need JSON5 fallback

Comment (rarkins) [reply lib/modules/manager/npm/post-update/npm.ts]: switched to zod

**Review by viceice — APPROVED** [review]

**Pull request merged**